### PR TITLE
rpc v2: `submitAndWatch` replace old messages if it's lagging

### PR DESCRIPTION
**Pull request by niklasad1**
Close https://github.com/paritytech/polkadot-sdk/issues/3076

The fix is really just that older messages are replaced if the client can't keep up with the server instead.
Because I wanted the same functionality as `pipe_from_stream` for both pending/subscription I added two wrapper types on-top of the types from jsonrpsee to make it nicer.

I added a trait `Buffer` so I could still use pipe_from_stream but that abstraction is a little leaky but only to avoid adding an identical method/function with another strategy...

### CODE REVIEW

File: substrate/client/rpc-spec-v2/src/transaction/transaction.rs [hunk 1]
```diff
@@ -108,16 +106,23 @@ where
...
+
 			match submit.await {
 				Ok(stream) => {
-					let stream = stream.filter_map(move |event| async move { handle_event(event) });
-					pipe_from_stream(pending, stream.boxed()).await;
+					let stream =
+						stream.filter_map(move |event| async move { handle_event(event) }).boxed();
+
+					// If the subscription is too slow older events will be overwritten.
+					sink.pipe_from_stream(stream.boxed(), RingBuffer::new(3)).await;
```
Comment (niklasad1) [inline substrate/client/rpc-spec-v2/src/transaction/transaction.rs:119]: This is the fix....

File: substrate/client/rpc-spec-v2/src/transaction/tests/transaction_tests.rs [hunk 2]
```diff
@@ -149,3 +149,89 @@ async fn tx_with_pruned_best_block() {
...
+	let block_1_header = api.push_block(1, vec![], true);
+	client.set_best_block(block_1_header.hash(), 1);
+
+	let uxt = uxt(Alice, ALICE_NONCE);
+	let xt = hex_string(&uxt.encode());
+
+	// The subscription itself has a buffer of length 1 and no way to create
+	// it without a buffer.
+	//
+	// Then `transactionWatch` has its own buffer of length 1 which leads to
```
Comment (niklasad1) [inline substrate/client/rpc-spec-v2/src/transaction/tests/transaction_tests.rs:165]:
```suggestion
	// Then `transactionWatch` has its own buffer of length 3 which leads to
```

File: substrate/client/rpc/src/utils.rs [hunk 3]
```diff
@@ -23,25 +23,52 @@ use futures::{
...
+		Self { inner: VecDeque::with_capacity(DEFAULT_BUF_SIZE), max_cap: DEFAULT_BUF_SIZE }
+	}
+}
+
 impl<T> BoundedVecDeque<T> {
 	/// Create a new bounded VecDeque.
-	fn new() -> Self {
-		Self { inner: VecDeque::with_capacity(DEFAULT_BUF_SIZE), max_cap: DEFAULT_BUF_SIZE }
+	pub fn new(cap: usize) -> Self {
+		Self { inner: VecDeque::with_capacity(cap), max_cap: DEFAULT_BUF_SIZE }
```
Comment (lexnv) [inline substrate/client/rpc/src/utils.rs:64]: Could the user pass a capacity higher than 16?

File: substrate/client/rpc/src/utils.rs [hunk 4]
```diff
@@ -50,126 +77,187 @@ impl<T> BoundedVecDeque<T> {
...
-///
-/// # Panics
-///
-/// This function panics `Serialize` fails and it is treated a bug.
-pub fn to_sub_message(sink: &SubscriptionSink, result: &impl Serialize) -> SubscriptionMessage {
-	SubscriptionMessage::new(sink.method_name(), sink.subscription_id(), result)
-		.expect("Serialize infallible; qed")
+	/// Send a message on the subscription.
+	pub async fn send(&self, result: &impl Serialize) -> Result<(), DisconnectError> {
+		self.0.send(self.to_sub_message(result)).await
```
Comment (lexnv) [inline substrate/client/rpc/src/utils.rs:243]: Nice! This makes the api easier to use!

**Review by lexnv — APPROVED** [review]
LGTM! Nice job here 👍

File: substrate/client/rpc/src/utils.rs (hunk 3 above)
Comment (niklasad1) [inline substrate/client/rpc/src/utils.rs:64]:
```suggestion
		Self { inner: VecDeque::with_capacity(cap), max_cap: cap }
```

**Review by pkhry — APPROVED** [review]

File: substrate/client/rpc/src/utils.rs [hunk 5]
```diff
@@ -50,126 +77,187 @@ impl<T> BoundedVecDeque<T> {
...
+
+		// Poll the stream while waiting for the subscription to be accepted
+		//
+		// If the `max_cap` is exceeded then the subscription is dropped.
+		let sink = loop {
+			match future::select(accept_fut, stream.next()).await {
+				Either::Left((Ok(sink), _)) => break sink,
+				Either::Right((Some(msg), f)) => {
+					if buf.push(msg).is_err() {
+						log::warn!(target: "rpc", "Subscription::accept buffer full for subscription={method} conn_id={conn_id}; dropping subscription");
```
Comment (niklasad1) [inline substrate/client/rpc/src/utils.rs:149]:
```suggestion
						log::debug!(target: "rpc", "Subscription::accept buffer full for subscription={method} conn_id={conn_id}; dropping subscription");
```

File: substrate/client/rpc/src/utils.rs [hunk 6]
```diff
@@ -50,126 +77,187 @@ impl<T> BoundedVecDeque<T> {
...
+				// Send operation finished.
+				Either::Right((Either::Left((_, n)), c)) => {
+					next_item = n;
+					closed = c;
+					next_fut = Box::pin(Fuse::terminated());
+				},
+				// New item from the stream
+				Either::Right((Either::Right((Some(v), n)), c)) => {
+					if buf.push(v).is_err() {
+						log::warn!(
```
Comment (niklasad1) [inline substrate/client/rpc/src/utils.rs:206]:
```suggestion
						log::debug!(
```

**Pull request merged**